src/core/meshcop/dtls.cpp

```diff
@@ -0,0 +1,378 @@
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file implements the necessary hooks for mbedTLS.
+ */
+
+#include <assert.h>
+#include <stdio.h>
+
+#include <common/code_utils.hpp>
+#include <common/encoding.hpp>
+#include <common/logging.hpp>
+#include <common/timer.hpp>
+#include <meshcop/dtls.hpp>
+#include <thread/thread_netif.hpp>
+
+#include <mbedtls/debug.h>
+
+namespace Thread {
+namespace MeshCoP {
+
+Dtls::Dtls(ThreadNetif &aNetif):
+    mStarted(false),
+    mTimer(&HandleTimer, this),
+    mTimerIntermediate(0),
+    mTimerSet(false),
+    mNetif(aNetif)
+{
+}
+
+ThreadError Dtls::Start(bool aClient, ReceiveHandler aReceiveHandler, SendHandler aSendHandler, void *aContext)
+{
+    static const int ciphersuites[2] = {0xC0FF, 0}; // EC-JPAKE cipher suite
+    int rval;
+
+    mReceiveHandler = aReceiveHandler;
+    mSendHandler = aSendHandler;
+    mContext = aContext;
+    mClient = aClient;
+
+    mbedtls_ssl_init(&mSsl);
+    mbedtls_ssl_config_init(&mConf);
+    mbedtls_ctr_drbg_init(&mCtrDrbg);
+    mbedtls_entropy_init(&mEntropy);
+
+    mbedtls_debug_set_threshold(4);
+
+    // XXX: should set personalization data to hardware address
+    rval = mbedtls_ctr_drbg_seed(&mCtrDrbg, mbedtls_entropy_func, &mEntropy, NULL, 0);
+    VerifyOrExit(rval == 0, ;);
+
+    rval = mbedtls_ssl_config_defaults(&mConf, mClient ? MBEDTLS_SSL_IS_CLIENT : MBEDTLS_SSL_IS_SERVER,
+                                       MBEDTLS_SSL_TRANSPORT_DATAGRAM, MBEDTLS_SSL_PRESET_DEFAULT);
+    VerifyOrExit(rval == 0, ;);
+
+    mbedtls_ssl_conf_rng(&mConf, mbedtls_ctr_drbg_random, &mCtrDrbg);
+    mbedtls_ssl_conf_min_version(&mConf, MBEDTLS_SSL_MAJOR_VERSION_3, MBEDTLS_SSL_MINOR_VERSION_3);
+    mbedtls_ssl_conf_max_version(&mConf, MBEDTLS_SSL_MAJOR_VERSION_3, MBEDTLS_SSL_MINOR_VERSION_3);
+    mbedtls_ssl_conf_ciphersuites(&mConf, ciphersuites);
+    mbedtls_ssl_conf_export_keys_cb(&mConf, HandleMbedtlsExportKeys, this);
+    mbedtls_ssl_conf_handshake_timeout(&mConf, 8000, 60000);
```

Comment:
Suggest that we change min timeout from 8000 ms to 20000 ms (20 seconds).
Max can be 100000 ms, if we want to max 5 attempts.